pkg/build/strategies/sti/sti.go

```diff
@@ -196,17 +208,24 @@ func (builder *STI) Build(config *api.Config) (*api.Result, error) {
 		glog.V(1).Infof("Running %q in %q", api.Assemble, config.Tag)
 	}
 	if err := builder.scripts.Execute(api.Assemble, config.AssembleUser, config); err != nil {
+		builder.result.BuildInfo = updater.UpdateFailureReason(builder.result.BuildInfo, api.ReasonAssembleFailed)
+
 		switch e := err.(type) {
 		case errors.ContainerError:
 			if !isMissingRequirements(e.Output) {
-				return nil, err
+				builder.result.BuildInfo = updater.UpdateFailureReason(builder.result.BuildInfo, api.ReasonUnmentDependenciesFailed)
+				return builder.result, err
 			}
 			glog.V(1).Info("Image is missing basic requirements (sh or tar), layered build will be performed")
+			// NOTE: separation of concerns, if we fail here, we shouldn't try
+			// another strategy here, but somewhere up the stack. In other words, a
+			// build() shouldn't call another build().
```

Comment:
debatable.  what you've got here is a uber-build strategy that understands how to delegate down to other build strategies if it can't do the build itself, and the uber-build understands the overall current state of the system (has already cloned the source, etc)